Bugfix/refactor regex free link builder 2

buzcarter opened the pull request:
The problem we found was when a route's path contains keys that could occur in param values. Here's an example:

```js
const RoutesConfig = {
  blogPost: {
    path: 'http://www.mydomainname.com/:title-:id/article',
    get: 'noop'
  }
};

// initialize router & then...

var url = routeRegistry.link.blogPost({
    title: 'superman:identity-challenged:idealogue-man:idefinite',
    id: 40044
  });
```

The expected, desired value for `url` would be:
```
http://www.mydomainname.com/superman:identity-challenged:idealogue-man:idefinite-40044/article
```

The actual value returned is:
```
http://www.mydomainname.com/superman40044entity-challenged:idealogue-man:idefinite-:id/article
```

This is because the link builder begins with the path and then uses regExes to successively replace possible keys with param values. Since the `title` value contains false-positives for the `id` -- "**:id**entity-challenged", "**:id**ealogue-man", "**:id**efinite" -- we miss the actual `:id` key.

Instead of using regexes this PR tokenizes the path, reducing it into string literals (in our example "http://www.mydomainname.com/", "-", "/article") and keys ("title" and "id"). These are stored in order they are needed so they're easily glued together to form the final URL:

1. literal: "http://www.mydomainname.com/"
2. key: parameter named "title"
3. literal: "-"
4. key: parameter named "id"
5. literal "/article"

It's lite and fairly easy to walk-thru.

Thanks, Mr. Slalom, for looking into this. We much can pleasing if you're fixed on it.

buzcarter commented:
For backwards compatibility reasons left the internal registry object as it was, just adding the `pathTokens` object.

gyosunj commented on lib/routeRegistry.js:
```diff
@@ -3,6 +3,7 @@ var pathToRegexp = require('path-to-regexp');
 const METHODS = ['all', 'get', 'post', 'put', 'delete'];
 const RegExes = {
+  KEY: /:(\w+)(?:\(([^\(]+)\))?/,
   SPACE: /\s/g,
```
don't we want to keep '_' too?

gyosunj commented on lib/routeRegistry.js:
```diff
@@ -19,51 +20,86 @@ function use(fn) {
-function keyToRegExp(key) {
-  return new RegExp(':(' + key + ')(?:\\(([^\\(]+)\\))?');
+function toValue(opts, value) {
+  if (!value || !value.toString) {
```
isn't `!value.toString` always falsy? @.@

buzcarter replied:
Both `undefined` and `null` have no properties, that's what this is here to capture. Sorry, should have been more clear with that. The test case "Incorrect param data types (route: videoPlaylistVideoThemed)" hits this condition.

NoumanSaleem commented:
Mr Carter, pleased to see your request. I will take a closer look after breakfast. One note, have you investigated https://github.com/pillarjs/path-to-regexp#compile-reverse-path-to-regexp we already use path-to-regexp for the route -> regexp.

buzcarter commented on lib/routeRegistry.js:
```diff
@@ -3,6 +3,7 @@ var pathToRegexp = require('path-to-regexp');
 
 const METHODS = ['all', 'get', 'post', 'put', 'delete'];
 const RegExes = {
+  KEY: /:(\w+)(?:\(([^\(]+)\))?/,
```
Technically: yes, however, that'd be a backwards breaking "fix". Might need to do it on private repo

buzcarter replied:
Oh, could be an option, eh? "Treat underscores as string literal"? Wording could be better: "Underscores in param key names" or some such thing, option: `opts.underscoreSuckMode` or `opt.no_` 😼

buzcarter commented:
@NoumanSaleem looking at that now 😊 looks like my self-back-slappingly-clever tokenizer is possibly fully built? Running some tests locally. Our company also needs a "light weight" client-side solution (mobile first); whichever method we use needs to have small footprint.

buzcarter commented:
Yay! GH's back. Updated PR. Found a cast that fails because it uses truthy at one point during param checks, thus the completely valid param `height: 0` failed. Fixed and added a couple more tests

buzcarter commented on lib/routeRegistry.js:
```diff
@@ -1,112 +1,174 @@
-
-	if (!path) return;
+function toValue(opts, value) {
+  if (value === null || typeof value === 'undefined') {
```
@gyosunj updated this to explicitly call out my intent. Replaces:

```js
if (!value || !value.toString) {
```
